riscv/interactive.cc

```diff
@@ -426,3 +628,248 @@ void sim_t::interactive_until(const std::string& cmd, const std::vector<std::str
+  if (r == NXPR) {
+    char *ptr;
+    r = strtoul(args[1].c_str(), &ptr, 10);
+    if (*ptr) {
+      #define DECLARE_CSR(name, number) if (args[1] == #name) csr_index = number;
+      #include "encoding.h"              // generates if's for all csrs
+      r = NXPR;                          // else case (csr name not found)
+      #undef DECLARE_CSR
+    }
+  }
+  if (r >= NXPR) {
+    if(csr_index == 0x1000)
+      throw trap_interactive();
+    else {
+      p->set_csr(csr_index, val);
+      return p->get_csr(csr_index);
+    }
+  }
+
+  p->get_state()->XPR.write(r, val);
+  return p->get_state()->XPR[r];
+}
+
+reg_t sim_t::set_mem(const std::vector<std::string>& args, reg_t val)
+{
+  if(args.size() != 1 && args.size() != 2)
+    throw trap_interactive();
+
+  std::string addr_str = args[0];
+  mmu_t* mmu = debug_mmu;
+  if(args.size() == 2)
+  {
+    processor_t *p = get_core(args[0]);
+    mmu = p->get_mmu();
+    addr_str = args[1];
+  }
+
+  reg_t addr = strtol(addr_str.c_str(),NULL,16), val2;
+  if(addr == LONG_MAX)
+    addr = strtoul(addr_str.c_str(),NULL,16);
+
+  switch(addr % 8)
+  {
+    case 0:
+      mmu->store_uint64(addr, val);
+      val2 = mmu->load_uint64(addr);
+      break;
+    case 4:
+      mmu->store_uint32(addr, (uint32_t)(val & (((reg_t)1 << 32) - 1)));
+      val2 = mmu->load_uint32(addr);
+      break;
+    case 2:
+    case 6:
+      mmu->store_uint16(addr, (uint16_t)(val & (((reg_t)1 << 16) - 1)));
+      val2 = mmu->load_uint16(addr);
+      break;
+    default:
+      mmu->store_uint8(addr, (uint8_t)(val & (((reg_t)1 << 8) - 1)));
+      val2 = mmu->load_uint8(addr);
+      break;
+  }
+  return val2;
+}
+
+reg_t sim_t::set_pc_value(const std::vector<std::string>& args, reg_t val)
+{
+  if(args.size() != 1)
+    throw trap_interactive();
+
+  processor_t *p = get_core(args[0]);
+
+  p->get_state()->pc = val;
+  return p->get_state()->pc;
+}
+
+int sim_t::get_reg_index(const std::vector<std::string>& args)
+{
+  if(args.size() != 2)
+    throw trap_interactive();
+
+  processor_t *p = get_core(args[0]);
+  int r = std::find(xpr_name, xpr_name + NXPR, args[1]) - xpr_name;
+
+  if (r == NXPR)
+    r = atoi(args[1].c_str());
+  if (!r)
+    return NXPR;
+  if (r >= NXPR)
+    throw trap_interactive();
+
+  return r;
+}
+
+void sim_t::interactive_history(const std::string& cmd, const std::vector<std::string>& args)
+{
+	for(int i = 0; i < history_list.size(); i++)
+		fprintf(stderr, "%-4d%s\n", i+1, history_list[i].c_str());
+}
+
+void sim_t::interactive_csr(const std::string& cmd, const std::vector<std::string>& args)
+{
+  if(args.size() != 1)
+    throw trap_interactive();
+
+    int i = 0;
+    std::map<std::string,reg_t> csr;
+    processor_t *p = get_core(args[0]);
+
+    #define CSR_DECLARE(name, number) csr[#name] = p->get_csr(number);
+    #include "encoding.h"              // generates if's for all csrs
+    #undef CSR_DECLARE
```

Comment:
Maybe using `DECLARE_CSR` would be better? and move `csr` to member data and then init once at constructor of `sim_t` or lazy init could prevent construct that again and again.

Reply:
thanks for review, i have modified it